### ISSUE TITLE
[HOLD for payment 2023-07-14] [$1000] Placeholder not moving back to it's place after selecting a contact

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#19231]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Click on FAB
2. Click on "Split bill"
3. Type an email and select the contact that shows up in the below suggestion
4. Notice the placeholder text "Name, email or phone number"

## Expected Result:
The placeholder text should return back to it's place after user selects on a contact

## Actual Result:
Placeholder stays in typing state and doesn't return back

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.16.3
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/9f767ec5-2b7d-4295-9d53-cd50a5e1f51f

https://github.com/Expensify/App/assets/43996225/802d475f-c737-4743-b533-30b44ccff1ec

**Expensify/Expensify Issue URL:**
**Issue reported by:** @nathan-mulugeta
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1684261779675679

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0196a89d054bb9c9e0</li>
        <li>Upwork Job ID: 1660771960364109824</li>
        <li>Last Price Increase: 2023-06-12</li>
    </ul>
</details>

**Comment by PrashantMangukiya** [comment]
## Proposal

Posting proposal early as per new [guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#propose-a-solution-for-the-job)

### Please re-state the problem that we are trying to solve in this issue.
Placeholder not moving back to it's place after selecting a contact. i.e. Placeholder not focus after selecting contact.

### What is the root cause of that problem?
During Split Bill it it rendering participant selection list via this
https://github.com/Expensify/App/blob/78ec5316c36c380ab64041a4b87144ae02cf80f3/src/pages/iou/steps/MoneyRequstParticipantsPage/MoneyRequestParticipantsSplitSelector.js#L218-L233

`OptionsSelector` has `shouldFocusOnSelectRow` prop is there with default value false. It can be used to set focus to text input after an option is selected. 
https://github.com/Expensify/App/blob/78ec5316c36c380ab64041a4b87144ae02cf80f3/src/components/OptionsSelector/optionsSelectorPropTypes.js#L71-L72

Within `MoneyRequestParticipantsSplitSelector.js` we are not passing `shouldFocusOnSelectRow` props to `OptionSelector` at line 218, so it will not set focus to text input after option is selected. This is the root cause of the problem.

### What changes do you think we should make in order to solve the problem?
Within `MoneyRequestParticipantsSplitSelector.js` we have to pass `shouldFocusOnSelectRow` to `OptionSelector` (starts from line 218) as shown code below. So it will focus to text input after option is selected.
```javascript
 <OptionsSelector 
     ....
     shouldFocusOnSelectRow        // *** Add this
 /> 
```
This will solve the issue as shown in result video.

### What alternative solutions did you explore? (Optional)
None

<details>
<summary>Result</summary>

https://github.com/Expensify/App/assets/7823358/83678391-fb6f-4a56-a9d3-f9418cbb932b

</details>

**Comment by strepanier03** [comment]
I am able to recreate this behavior, but something odd also happened. 

When testing, I recreated the behavior right away. But then I stepped away and left my screen open on the contact select screen where this issue was happening. 

When I came back 5-10 minutes later, the behavior didn't happen any longer. The placeholder went back to it's place every time and very quickly. Now, I can't recreate the issue again. 

I'm going to pause until tomorrow and then move this forward then if I can recreate.

**Comment by Nathan-Mulugeta** [comment]
Yes @strepanier03, I am still able to reproduce this. Here are some things to notice. After clicking split bill from the FAB menu instead of selecting the contact lists from the recent list, type out a new email to add to the split bill money request. After typing the email in the input field select the email that shows up. Then you can notice the place holder doesn't return back to its place. I will attach a video demonestrating this.   

https://github.com/Expensify/App/assets/111440031/57b02fa2-645b-493c-80dc-cabfeda6f801

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Text input label does not deactivate (move back as the placeholder) when we clear the input value. To reproduce it, make sure type something first.

### What is the root cause of that problem?
Actually, we already have the code that SHOULD deactivate the label when we clear the input value here:
https://github.com/Expensify/App/blob/737862677811acc2b05c68209a56ecd58bee59f3/src/components/TextInput/BaseTextInput.js#L71-L89

As you can see, we update the `value` state with `inputValue` and call `deactivateLabel` if `inputValue` is empty. Then, inside `deactivateLabel`, we check if the state of `value` is empty or not. 
https://github.com/Expensify/App/blob/737862677811acc2b05c68209a56ecd58bee59f3/src/components/TextInput/BaseTextInput.js#L164-L171

If it's empty, then we proceed to deactivate the label. The problem is, the `value` state is not updated yet inside `deactivateLabel`, which is basically a race condition.

### What changes do you think we should make in order to solve the problem?

Call the `deactivateLabel` inside the `setState` callback.

More specific, move this code to be inside the `setState` callback.
https://github.com/Expensify/App/blob/737862677811acc2b05c68209a56ecd58bee59f3/src/components/TextInput/BaseTextInput.js#L85-L89

I guess we also can change `inputValue` to `this.state.value` instead.

**Result**

https://github.com/Expensify/App/assets/50919443/1ddc61e5-9f65-465c-a202-9fe608ea0f3e

**Comment by strepanier03** [comment]
Thank you for the extra details @Nathan-Mulugeta! Moved this on.

**Comment by dukenv0307** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Placeholder stays in typing state and doesn't return back after selecting a contact.

### What is the root cause of that problem?
There're 2 problems:

1. First, the expected result might not be enough. Here's what happens when a contact is selected in Android/iOS:

https://github.com/Expensify/App/assets/129500732/cf9ff4a8-c65d-483e-81a1-47479a0c025f

As we can see, on native platforms, when a contact is selected, the search field will remain focused. The same should happen for web as well for consistency, while currently for web the search input will be blurred. Also there's no reason we should blur the input after selecting a contact, because the user might want to search and add a new contact immediately.

Also worthy to note that `the search field will remain focused` is different from `the search field will be blurred then focused again` which is what the `shouldFocusOnSelectRow` is for. `the search field will be blurred then focused again` will make the input blink when selecting contact, also it's different UX compared to Android/iOS.

So the expectation should be "the search input should remain focused on Web after selecting a contact" (currently it's like that for Android/iOS).

The root cause is that currently we're not supporting this behavior.

2. `1.` should be enough to fix this issue, but there's a small issue in here https://github.com/Expensify/App/blob/1aa30d1ecfa8a69275b49f051ad65c392ae4e985/src/components/TextInput/BaseTextInput.js#L88 which might cause problems in the future. In there the `deactivateLabel` is called if the `inputValue` becomes empty, inside `deactivateLabel` itself we're using the `state.value` here https://github.com/Expensify/App/blob/1aa30d1ecfa8a69275b49f051ad65c392ae4e985/src/components/TextInput/BaseTextInput.js#L165. It causes race condition where `inputValue` becomes empty but the `state.value` doesn't update to empty yet inside `deactivateLabel`, causing the label not to be deactivated and moved back to its place properly.

### What changes do you think we should make in order to solve the problem?


1.
We need to listen to `mouse down` event of the option, and `preventDefault` if the `TextInput` is currently focused, so that it will not blur the `TextInput` on selecting option.

We already use the same approach to make sure the Composer is not blurred when pressing the send button [here](https://github.com/Expensify/App/blob/1ac9964003fcb61d6107c205ce7573d67e170ec3/src/pages/home/report/ReportActionCompose.js#L1107).

Steps:
a. Pass a new `onRowMouseDown` to `<OptionsList` [here](https://github.com/Expensify/App/blob/1ac9964003fcb61d6107c205ce7573d67e170ec3/src/components/OptionsSelector/BaseOptionsSelector.js#L301)

```
onRowMouseDown={(e) => {
    if (!e || !this.textInput.isFocused()) {
        return;
    }
    
    e.preventDefault();
}}
``` 

b. Pass `onRowMouseDown={this.props.onRowMouseDown}` to `OptionRow`, then in `OptionRow` here https://github.com/Expensify/App/blob/1ac9964003fcb61d6107c205ce7573d67e170ec3/src/components/OptionRow.js#L151, assign the `onRowMouseDown` to the `onMouseDown`:

```onMouseDown={this.props.onRowMouseDown}```

2. To fix this we should make sure the `deactivateLabel` call [here](https://github.com/Expensify/App/blob/1aa30d1ecfa8a69275b49f051ad65c392ae4e985/src/components/TextInput/BaseTextInput.js#L88) happens inside [this setState callback](https://github.com/Expensify/App/blob/1aa30d1ecfa8a69275b49f051ad65c392ae4e985/src/components/TextInput/BaseTextInput.js#L78)

### What alternative solutions did you explore? (Optional)
In `1`, we can optionally introduce a variable to `OptionsSelector` to control if we want to disable the behavior `"the search input should remain focused on Web after selecting a contact"` in some cases

**Comment by strepanier03** [comment]
@sobitneupane - Friendly bump for the proposal submitted yesterday. Thank you!

**Comment by sobitneupane** [comment]
I believe the behavior of `OptionsSelector` in split bill page should be similar to that of `OptionSelector` in New group page.

https://github.com/Expensify/App/assets/25876548/0cf71b83-fee5-4d61-af12-fe4171f29ebd

In New group page,

- Text Input refocuses on select row with input text selected.


@strepanier03 @tylerkaraszewski Do we want similar behavior in Split bill page as well?

**Comment by dukenv0307** [comment]
@sobitneupane @tylerkaraszewski just to note both `Split bill` page and `New group` page behavior after selecting the option in desktop are inconsistent with that same page in mobile. In mobile the input will not be blurred (blinked) after the option is selected. We can make all those pages consistent with each other and also platforms to be consistent as well.

**Comment by strepanier03** [comment]
@dukenv0307 - Looks like Sobit left a comment on your proposal.

**Comment by sobitneupane** [comment]
@strepanier03 @tylerkaraszewski Do you agree on [these points](https://github.com/Expensify/App/issues/19231#issuecomment-1562684702)?

**Comment by bernhardoj** [comment]
Giving my opinion.

I think it's better for us to solve the initial reported bug first, rather than covering it up by refocusing the text input.

**Comment by strepanier03** [comment]
I think we need to raise [this](https://github.com/Expensify/App/issues/19231#issuecomment-1562684702) for discussion on the best path forward so I'll do that first thing tomorrow when more of the team is online.

**Comment by tylerkaraszewski** [comment]
@strepanier03 - do you have a slack thread or something with that discussion in it?

**Comment by strepanier03** [comment]
@tylerkaraszewski - Oh damn, this fell through the cracks last week and I dropped the ball here. So sorry about that. 

I started the discussion [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1686001589352959). Thank you for the bump 🙌

**Comment by strepanier03** [comment]
Bumped the discussion in #expensify-open-source [again](https://expensify.slack.com/archives/C01GTK53T8Q/p1686179932337289).

**Comment by sobitneupane** [comment]
Thanks @strepanier03.

@dukenv0307 Would you like to update your proposal to match behavior in New group page. [This](https://github.com/Expensify/App/issues/19231#issuecomment-1562684702) is what happens in new group page.
>In New group page,
> Text Input refocuses on select row with input text selected.

**Comment by strepanier03** [comment]
@dukenv0307 - Thanks for posting on the thread and moving things forward, I appreciate the knowledge you brought to the discussion. 

Looks like your [suggestion](https://expensify.slack.com/archives/C01GTK53T8Q/p1686238266928599?thread_ts=1686179932.337289&cid=C01GTK53T8Q) has buy in though so I think we're good to move forward.

**Comment by tylerkaraszewski** [comment]
👍 go ahead.

**Comment by sobitneupane** [comment]
@dukenv0307 Does [your proposal](https://github.com/Expensify/App/issues/19231#issuecomment-1559698481) meet the latest expected output or do you want to update it?

**Comment by sobitneupane** [comment]
@dukenv0307 Similar to Create Group, both on selecting or deselecting a row, value input in the search field should be selected.

Overall, [Proposal](https://github.com/Expensify/App/issues/19231#issuecomment-1559698481) from @dukenv0307 looks good to me.

🎀👀🎀 C+ reviewed
cc: @tylerkaraszewski

**Comment by strepanier03** [comment]
@tylerkaraszewski - Friendly bump on reviewing the [proposal](https://github.com/Expensify/App/issues/19231#issuecomment-1559698481) that @sobitneupane thinks looks good.

**Comment by strepanier03** [comment]
@isabelastisser - I am heading out on sabbatical this evening and need to reassign this. There is no immediate action you need to take on this today, I bumped the eng for review and we just need to keep pressure on and this moving forward. 

Thank you for handling the wrap up of this one!

**Comment by isabelastisser** [comment]
Hey @tylerkaraszewski , can you please follow up on this? Thanks!

https://github.com/Expensify/App/issues/19231#issuecomment-1593636886

**Comment by isabelastisser** [comment]
Bump @tylerkaraszewski

**Comment by isabelastisser** [comment]
Assigning another team member because I will be OOO until July 5. 
[ SO](https://stackoverflowteams.com/c/expensify/questions/8836) for reference.

**Comment by alexpensify** [comment]
Waiting for a review here before assigning it out.

**Comment by alexpensify** [comment]
No update yet

**Comment by tylerkaraszewski** [comment]
Sorry about all the delays on this. Assigned now.

**Comment by alexpensify** [comment]
There we go, automation is in place now! @dukenv0307 the issue has been assigned to you.

**Comment by sobitneupane** [comment]
We are having [discussion](https://expensify.slack.com/archives/C01GTK53T8Q/p1687282549853709) on expected behavior on selecting and unselecting row on New Group and Split Bill pages. So, putting hold on PR.

**Comment by alexpensify** [comment]
@sobitneupane should we put this GH on hold too?

**Comment by alexpensify** [comment]
@isabelastisser - I'm going to reassign another `bug` team member, I'm going OOO until Monday, July 10. I see you are returning on July 5. 

Required action from the 🐛 team: Check on the PR state and put it on Hold if we need more time. Maybe move to weekly.

**Comment by sobitneupane** [comment]
> should we put this GH on hold too?

@alexpensify I don't think that will be necessary. I believe we will come to conclusion within a day or two.

**Comment by alexpensify** [comment]
Thank you for clarifying @sobitneupane!

**Comment by tylerkaraszewski** [comment]
Hey @arosiclair - I am leaving on sabbatical so you got picked to take over this issue. I think I'll you'll need to do is review the final PR when it's ready.

**Comment by arosiclair** [comment]
@sobitneupane what's the status of this issue?

**Comment by sobitneupane** [comment]
PR is off hold now. Decision was made in the [thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1688147188152929?thread_ts=1687282549.853709&cid=C01GTK53T8Q) to highlight text on selecting/deselecting row.

**Comment by isabelastisser** [comment]
I'm back. Thanks @alexpensify!

**Comment by isabelastisser** [comment]
@dukenv0307 @sobitneupane did this PR cause a regression?

**Comment by sobitneupane** [comment]
Nope @isabelastisser The linked issue is not a bug. It is an expected behavior to focus the textinput on selecting/unselecting row.

**Comment by isabelastisser** [comment]
@sobitneupane, to confirm, the PR didn't cause a regression, and this issue is ready for payment, correct? 

@arosiclair I think we can process the bonus because the PR merged shortly after it was no longer on hold. Do you agree?

**Comment by sobitneupane** [comment]
@isabelastisser Yup. The PR didn't cause a regression. The PR has not hit the production yet. So, not ready to issue payment.

**Comment by arosiclair** [comment]
> @arosiclair I think we can process the bonus because the PR merged shortly after it was no longer on hold. Do you agree?

That makes sense to me 👍

**Comment by isabelastisser** [comment]
I was OOO yesterday and will take care of this by EOD.

**Comment by isabelastisser** [comment]
@sobitneupane please complete the BugZero checklist.

I will process the payments in Upwork now.

**Comment by sobitneupane** [comment]
Sorry for the delay @isabelastisser. I will complete the checklist by tomorrow. I will be paid through newDot.

**Comment by isabelastisser** [comment]
The payments were made in Upwork. Waiting for @sobitneupane to complete the BZ checklist, then we are all set.

**Comment by sobitneupane** [comment]
> **BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:
> 

Rather than a bug, it was an inconsistency issue. The PR raised make `MoneyRequestParticipantsSplitSelector ` page consistent with NewGroupPage.

**Comment by sobitneupane** [comment]
**Regression Test Proposal**

- Click on FAB
- Click on "Split bill"
- Type an email and select the contact that shows up in the below suggestion
- Verify that the search input is highlighted/selected after user selects/deselects

Do we agree 👍 or 👎

**Comment by JmillsExpensify** [comment]
@isabelastisser Can you please summarize the appropriate individual payments for all parties involved in this issue? This is holding up @sobitneupane's NewDot payments. More information on this compliance process [in Slack](https://expensify.slack.com/archives/C01SKUP7QR0/p1690321157273869).

**Comment by isabelastisser** [comment]
Payments summary:

Issue reported by: @Nathan-Mulugeta $250
C+ review: @sobitneupane $1000 review, $500 urgency bonus
Issue fixed by: @dukenv0307, $1000 for accepted proposal, $500 urgency bonus

**Comment by JmillsExpensify** [comment]
Reviewed details for @sobitneupane. These details are accurate based on summary from Business Reviewer and are now approved for payment in NewDot.
